plugins/contiv/ipam/ipam.go

```diff
@@ -243,29 +246,50 @@ func (i *IPAM) NextPodIP(podID string) (net.IP, error) {
 		return nil, err
 	}
 
+	last := i.lastAssigned + 1
 	// iterate over all possible IP addresses for pod network prefix
-	// and take first not assigned IP
+	// start from the last assigned and take first available IP
 	prefixBits, totalBits := i.podNetworkIPPrefix.Mask.Size()
 	maxSeqID := 1 << uint(totalBits-prefixBits) //max IP addresses in network range
-	for j := 1; j < maxSeqID; j++ {             // zero ending IP is reserved for network => skip seqID=0
-		if j == podGatewaySeqID {
-			continue // gateway IP address can't be assigned as pod
-		}
-		if _, found := i.assignedPodIPs[networkPrefix+uint32(j)]; found {
-			continue // ignore already assigned IP addresses
+	for j := last; j < maxSeqID; j++ {          // zero ending IP is reserved for network => skip seqID=0
+		ipForAssign, success := i.tryToAllocatePodIP(j, networkPrefix, podID)
+		if success {
+			i.lastAssigned = j
+			return ipForAssign, nil
 		}
-		i.assignedPodIPs[networkPrefix+uint32(j)] = podID
-		//TODO set etcd for new assigned value
+	}
 
-		ipForAssign := uint32ToIpv4(networkPrefix + uint32(j))
-		i.logger.Infof("Assigned new pod IP %s", ipForAssign)
-		i.logAssignedPodIPPool()
-		return ipForAssign, nil
+	// iterate from the range start until lastAssigned
+	for j := 1; j < last; j++ { // zero ending IP is reserved for network => skip seqID=0
```

Comment:
you might want to wrap the try assign as a method or just put it as part of `tryToAllocatePodIP`, so that in case we want to change the algorithm again.